proposals/0046-actions.md

```diff
@@ -0,0 +1,637 @@
+- **You can call an action using a standard HTML `form` element**.
+- **You can use client JavaScript to call an action** using scripts or islands. When doing so, data returned by the action is type-safe without type casting. Note: This should consider form handling in popular component frameworks, like the [`useActionState()`](https://react.dev/reference/react-dom/hooks/useFormState) and [`useFormStatus()`](https://react.dev/reference/react-dom/hooks/useFormStatus) hooks in React 19.
+- **You can declare actions as an endpoint** to call from prerendered / static pages.
+
+# Non-Goals
+
+- **A solution to client-side validation.** Validation is a major piece to forms with several community libraries to choose from (ex. react-hook-form). Astro may recommend standard HTML attributes for client validation including the `required` and `type` properties on an input.
+- Declaring actions within `.astro` frontmatter. Frontmatter forms a function closure, which can lead to misuse of variables within an action handler. This challenge is shared by `getStaticPaths()` and it would be best to avoid repeating this pattern in future APIs.
+
+# Detailed Design
+
+All actions are declared in a global actions handler: the `src/actions.ts` file. Similar to Astro middleware, users may switch to using `src/actions/index.ts` to store related code in a directory.
+
+You can define an action using the `defineAction()` utility from the `astro:actions` module. This accepts the `handler` property to define your server-side request handler. If your action accepts arguments, apply the `schema` property to validate parameters with Zod.
+
+This example defines a `like` action, which accepts a `postId` as a string and updates the like count on a related database entry. That action is exposed for use in your application with an exported `server` object passing `like` as a property:
+
+```ts
+// src/actions/index.ts
+import { defineAction, z } from "astro:actions";
+import { db, Likes, eq, sql } from "astro:db";
+
+export const server = {
+  like: defineAction({
+    input: z.object({ postId: z.string() }),
+    handler: async ({ postid }) => {
+      const { likes } = await db
+        .update(Likes)
+        .set({
+          likes: sql`likes + 1`,
+        })
+        .where(eq(Likes.postId, postId))
+        .returning()
+        .get();
+      return likes;
+    },
+  }),
+};
+```
+
+Now, this action is callable from client components and server forms.
+
+> **@bholmesdev Note:** Users will call actions by importing an `actions` object from the `astro:actions` virtual module. To avoid misleading auto-import suggestions, we use the name `server` instead of `actions` for server code.
```

Comment:
Would using the `default` export have this problem? I think `default` would be preferred for API consistency reasons, the Astro config is a default export, renderer APIs use a default export, etc. I think that Content Collections is the exception here.

Unless there's a problem with default exports I'm not aware of.

Reply:
I think avoiding auto-import collisions is a good reason, so I'm fine with sticking with `server` or coming up with another name.